### PR TITLE
implement node:dns module for nodejs_compat

### PR DESCRIPTION
**Pull request by anonrig**
Implements all functions in `node:dns` using Cloudflare DNS (1.1.1.1)

### CODE REVIEW

File: src/node/internal/internal_dns.ts
Comment (vicb) [inline src/node/internal/internal_dns.ts:75]: would value.data.slice(1, -1) be faster ?

Comment (anonrig) [reply src/node/internal/internal_dns.ts:75]: Nice catch. Yes, definitely!

File: src/node/internal/internal_dns_client.ts
Comment (anonrig) [reply src/node/internal/internal_dns_client.ts:52]: Yes, also I need to find a way to distinguish found/not found DNS records.

```
❯ curl --header "accept: application/dns-json" "https://cloudflare-dns.com/dns-query?name=google.com2222&type=TXT" | jq
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   246  100   246    0     0   5104      0 --:--:-- --:--:-- --:--:--  5020
{
  "Status": 3,
  "TC": false,
  "RD": true,
  "RA": true,
  "AD": true,
  "CD": false,
  "Question": [
    {
      "name": "google.com2222",
      "type": 16
    }
  ],
  "Authority": [
    {
      "name": "",
      "type": 6,
      "TTL": 86400,
      "data": "a.root-servers.net. nstld.verisign-grs.com. 2024112701 1800 900 604800 86400"
    }
  ]
}
```

File: src/workerd/api/node/tests/dns-nodejs-test.js
Comment (vicb) [inline src/workerd/api/node/tests/dns-nodejs-test.js:51]:
```suggestion
        results[0][0][0] !== '"',
```

File: src/workerd/api/node/tests/dns-nodejs-test.js
Comment (vicb) [inline src/workerd/api/node/tests/dns-nodejs-test.js:73]:
```suggestion
      // Callback API
      dns.resolveTxt('cloudflare.com', (error, results) => {
        if (error) {
          return fail(error);
        }
        testResults(results);
      });
```

**Review by vicb — COMMENTED** [review]
Added a few inline comments

File: src/node/dns.ts
Comment (jasnell) [inline src/node/dns.ts:46]: You know we implement `callbackify`? You can move its implementation into `node/internal` and reuse that here.

File: src/node/dns.ts
Comment (jasnell) [inline src/node/dns.ts:67]: Since you're going from promises to callbacks here, how are you handling synchronously throwing on argument validation? By default the JSG wrapper should cause synchronously thrown errors at the c++ level to be returned into promise rejections. Are you certain this will do the right thing for the errors the Node.js API expects to be thrown synchronously?

File: src/node/dns.ts
Comment (jasnell) [inline src/node/dns.ts:54]: Node.js will throw an error on these if `callback` is not a function, correct?

File: src/node/internal/internal_dns_client.ts
Comment (jasnell) [inline src/node/internal/internal_dns_client.ts:52]: Since this will route these requests as public internet requests I guess we should decide if using global fetch is really what we want vs. using a private fetch via a binding that would for sure route the request through internal infrastructure.

File: src/node/internal/internal_dns_client.ts
Comment (anonrig) [reply src/node/internal/internal_dns_client.ts:77]: On typescript `[0]` doesn't return `undefined | ValueType` but `.at()` does. I think we should encourage it more. And we should also enable `noUncheckedIndexedAccess` on tsconfig as well (ref: https://www.typescriptlang.org/tsconfig/#noUncheckedIndexedAccess)

File: src/node/internal/internal_errors.ts
Comment (anonrig) [reply src/node/internal/internal_errors.ts:601]: Yes

File: src/node/dns.ts
Comment (anonrig) [reply src/node/dns.ts:46]: TIL. Thanks

File: src/node/dns.ts
Comment (anonrig) [reply src/node/dns.ts:67]: I'll remove `async` from functions and just return a promise from the function, which would throw validation errors synchronously but the function will return Promise

File: src/node/dns.ts
Comment (anonrig) [reply src/node/dns.ts:54]: Yes, definitely there is a `validateFunction` missing from here.

File: src/node/internal/internal_dns_client.ts
Comment (anonrig) [reply src/node/internal/internal_dns_client.ts:61]: Is there an internal API to remove this sub request limit from this call?

File: src/node/internal/internal_dns_client.ts
Comment (anonrig) [reply src/node/internal/internal_dns_client.ts:52]: Yes, for sure, we should find a solution to this before landing this pull-request.

File: src/node/internal/internal_dns_client.ts
Comment (vicb) [reply src/node/internal/internal_dns_client.ts:77]: > On typescript [0] doesn't return undefined | ValueType but .at() does. 

Did you mean to say "does not"?

If that's the case TS is lying to you:

image

Comment (vicb) [reply src/node/internal/internal_dns_client.ts:77]: TS doesn't lie on my machine

image

File: src/workerd/api/node/tests/dns-nodejs-test.js
Comment (vicb) [reply src/workerd/api/node/tests/dns-nodejs-test.js:51]: See [the previous comment](https://github.com/cloudflare/workerd/pull/3183/files/19f1629ff3537d5b115a4f51c7196a63fc9a90ad#r1864151200)

Comment (vicb) [reply src/workerd/api/node/tests/dns-nodejs-test.js:51]: So I guess it should be `results[0]?.[0]?.[0]`

File: src/node/internal/internal_dns_client.ts
Comment (vicb) [reply src/node/internal/internal_dns_client.ts:77]: Both return `undefined | T`

File: src/workerd/api/node/tests/dns-nodejs-test.js
Comment (anonrig) [reply src/workerd/api/node/tests/dns-nodejs-test.js:73]: @jasnell Does this hold, or do we need Promise.withResolvers() to wait for the callback response?

File: src/node/internal/internal_dns_client.ts
Comment (anonrig) [reply src/node/internal/internal_dns_client.ts:77]: TS playground disagrees: https://www.typescriptlang.org/play/?target=99#code/MYewdgzgLgBAllApgWwgLhtATnMBzAbQF0YBeGAgcgAtEAbOkSgGhkoHcQs6ATSogFAC4AMxgAKBCggEAzEQIAGEqXIAiamoCUMAN4CYMUJBB1EAOkZ5xaqNTgQYPEIghhKsEQEM4dbQG4BAF8hUQkpVHMvKHFZLQB+KJjFLR19Q2MIUwsrGzsHJxc3DxhvX21g0LFJJEjo2K0k8RS0gyNwLLNLEGtbe0dnV3dYACMAV18eCqCgA

Comment (vicb) [reply src/node/internal/internal_dns_client.ts]: Oh 🙀
TIL!
Still you should (even more) add the "?" here and below!

File: src/node/dns.ts
Comment (jasnell) [inline src/node/dns.ts:51]: Need to verify that this does the right thing for argument validation.

File: src/node/internal/internal_dns.ts
Comment (jasnell) [inline src/node/internal/internal_dns.ts:50]: Why not async await syntax?

File: src/node/internal/internal_dns_client.ts
Comment (jasnell) [reply src/node/internal/internal_dns_client.ts:62]: Not using global fetch. If we were using a binding or exposing the fetcher a different way then yes. But honestly I think dns calls *should* be subject to the limit.

File: src/workerd/api/node/tests/dns-nodejs-test.js
Comment (jasnell) [reply src/workerd/api/node/tests/dns-nodejs-test.js]: The use of `Promise.withResolvers()` looks correct here.

File: src/node/internal/internal_dns.ts
Comment (anonrig) [reply src/node/internal/internal_dns.ts:50]: Because `validateString()` should throw synchronously whereas the DNS request errors should be async errors.

Comment (jasnell) [reply src/node/internal/internal_dns.ts:50]: A comment along those lines would be helpful

File: src/node/internal/internal_dns.ts
Comment (jasnell) [inline src/node/internal/internal_dns.ts:54]: To be extra safe here (in case the CF DNS behavior ever changed unexpectedly), it might be worth going ahead and checking to see if the answer endsWith "."

File: src/node/internal/internal_dns.ts
Comment (jasnell) [inline src/node/internal/internal_dns.ts:49]: To reduce code duplication, consider moving this check into a reusable utility function.

```js
function validateAnswer(json, name, op) {
  if (!('Answer' in json)) {
    throw new DnsError(name, errorCodes.NODATA, op);
  }
}
```

```
validateAnswer(json, name, 'queryCname');
validateAnswer(json, name, 'queryCaa');
// ... etc
```

File: src/node/internal/internal_dns.ts
Comment (jasnell) [inline src/node/internal/internal_dns.ts:63]: Nit: Might be a bit nicer to define a type for this.

File: src/node/internal/internal_dns.ts
Comment (jasnell) [inline src/node/internal/internal_dns.ts:108]: I assume this is also because the value may end in a "."? If so, a comment here would be helpful along with a double check to make sure the value ends with the "." before slicing it.

File: src/node/internal/internal_dns.ts
Comment (jasnell) [inline src/node/internal/internal_dns.ts:55]: I notice that Node.js does not actually enforce this. If `options` is anything other than an object Node.js appears to ignore it.

File: src/node/internal/internal_dns.ts
Comment (jasnell) [inline src/node/internal/internal_dns.ts:57]: Likewise here. Node.js appears to handle any truthy value for `options.ttl`

File: src/node/internal/internal_dns.ts
Comment (jasnell) [inline src/node/internal/internal_dns.ts:55]: Should be able to simplify to just....

```suggestion
  const ttl = !!options?.ttl;
```

**Comment by jasnell** [comment]
I think the only key remaining question here is whether this should be using a global fetch to the cloudflare dns endpoint or whether it should be using an internal route to DNS in production. 

@kentonv thoughts?

File: src/node/internal/internal_dns.ts
Comment (anonrig) [reply src/node/internal/internal_dns.ts:55]: I have how easy JS enables you to write safe JS code that looks unsafe.

File: src/node/internal/internal_errors.ts
Comment (vicb) [inline src/node/internal/internal_errors.ts:610]:
```suggestion
  constructor(public hostname: string, code: string, public syscall: string) {
    super(code, `${syscall} ${code} ${hostname}`);
  }
}
```

File: src/node/internal/internal_utils.ts
Comment (vicb) [inline src/node/internal/internal_utils.ts:54]: Why is this needed? enc is supposed to be a string.

Also `ucs2` is repeated l52 and l56

File: src/node/internal/internal_utils.ts
Comment (vicb) [inline src/node/internal/internal_utils.ts:60]: `${enc}.toLowerCase() === 'hex'` would only ever be usefull if enc is not a string. 
Maybe the type of enc needs to be fixed on that function?

File: src/node/internal/internal_utils.ts
Comment (vicb) [inline src/node/internal/internal_utils.ts:218]: would `String(reason)` allow removing the comment?

File: src/node/internal/internal_utils.ts
Comment (vicb) [inline src/node/internal/internal_utils.ts:365]:
```suggestion
  if (typeof descriptors.length?.value === 'number') {
    descriptors.length.value++;
  }
  if (typeof descriptors.name?.value === 'string') {
    descriptors.name.value += 'Callbackified';
  }
```

File: src/rust/dns/lib.rs
Comment (vicb) [inline src/rust/dns/lib.rs:89]: Nice :)

Is `- Replacement length` wrongly positioned?

**Review by vicb — COMMENTED** [review]
Great PR!
A few minor comments

File: src/node/internal/internal_dns.ts
Comment (anonrig) [reply src/node/internal/internal_dns.ts:70]: Performance wise this doesn't make a difference, but looks nicer, I agree.

File: src/node/internal/internal_dns_client.ts
Comment (anonrig) [reply src/node/internal/internal_dns_client.ts:101]: This will be slower, and hence we know that if it exists, it will be an array, we can just check for `null` or `undefined` value.

File: src/node/internal/internal_dns_client.ts
Comment (anonrig) [reply src/node/internal/internal_dns_client.ts:135]: There is no behavioral/performance impact of the recommendation, and I personally think it reduces the readability.

File: src/node/internal/internal_dns_client.ts
Comment (anonrig) [reply src/node/internal/internal_dns_client.ts:152]: This would be better:

```
const record = dnsUtil.parseCaaRecord(data);
  const obj: CAA = { critical: record.critical };
  obj[record.field] = record.value;
  return obj;
  ```

File: src/node/internal/internal_dns_client.ts
Comment (anonrig) [reply src/node/internal/internal_dns_client.ts:169]: I'll be using these functions in resolveAny method, where the separation of these functions make it much more readable.

File: src/node/internal/internal_utils.ts
Comment (anonrig) [reply src/node/internal/internal_utils.ts:365]: Yeah, this is better, but we should definitely add an eslint rule that forbids this. I'll apply this change but there is nothing blocking others from reverting this.

File: src/node/internal/internal_utils.ts
Comment (anonrig) [reply src/node/internal/internal_utils.ts:218]: yes. i personally think we should just remove this linting rule in the future.

**Comment by kentonv** [comment]
> I think the only key remaining question here is whether this should be using a global fetch to the cloudflare dns endpoint or whether it should be using an internal route to DNS in production.

It's important that these requests do not bypass egress workers, since that could be a security problem if the egress worker is intentionally blocking communications with the internet. If we had a DNS handler API, then we could let the egress worker implement that and proxy DNS requests explicitly -- but that would be a big project. So the only reasonable alternative right now is to go through global fetch() or global connect(). But we don't actually have a connect handler API yet either... so in that case requests would just fail in the presence of an egress worker. So global fetch() seems like the best option.

> Also, since this is a subrequest to a CF property, should the subrequest propagate tracing metadata, control header, etc.

I think that's a more general question -- should global fetch() to a CF property propagate tracing metadata etc.? I don't think that's something we need to answer in the context of this PR specifically.

File: src/node/internal/internal_dns.ts
Comment (anonrig) [reply src/node/internal/internal_dns.ts:51]: `!!options?.ttl` because node.js omits any non-object value and doesn't validate if TTL is a boolean

File: src/rust/dns/lib.rs
Comment (anonrig) [reply src/rust/dns/lib.rs:89]: Yes! Nice catch.

File: src/rust/dns/lib.rs
Comment (mikea) [inline src/rust/dns/lib.rs:26]: We shouldn't panic. Especially during parsing. Let's replace this with errors. I suggest using `thiserror` - very easy to use.

File: src/rust/dns/lib.rs
Comment (mikea) [inline src/rust/dns/lib.rs:19]: these should be Result. Parsing could fail by definition. Result is supported by cxx and is transformed into exception.

File: src/rust/dns/lib.rs
Comment (mikea) [inline src/rust/dns/lib.rs:64]: no panic please

File: src/rust/dns/lib.rs
Comment (mikea) [inline src/rust/dns/lib.rs:158]: maybe a simple comment about what's going on here? Hard to decode intent.

File: src/node/internal/internal_dns.ts
Comment (jasnell) [inline src/node/internal/internal_dns.ts:30]: A comment on these "Not implemented" stubs describing them a bit more would be helpful for future contributors. Are we not implementing them for a particular reason? Is there a path to implementing them? Links to the Node.js docs describing them, etc. Basically just something that would give enough context for someone coming into the project later to know what, if anything, to do with these.

File: src/node/internal/internal_dns.ts
Comment (jasnell) [inline src/node/internal/internal_dns.ts:48]: Non-blocking: Ugh I've always hated the fact that these returns are polymorphic

File: src/node/internal/internal_dns.ts
Comment (jasnell) [inline src/node/internal/internal_dns.ts:86]: A comment here especially would be helpful. Why are we not implementing the `resolveAny(...)` when we implement the rest of these.

**Review by jasnell — APPROVED** [review]
LGTM for everything except the Rust code, which I'll let someone else review. A couple nits regarding more code comments that I'd like to see addressed before this lands but otherwise the code good.

**Review by jasnell — APPROVED** [review]

File: src/rust/dns/lib.rs
Comment (mikea) [inline src/rust/dns/lib.rs:40]: I personally don't care about errors documentation much (panics are important). Is clippy requiring them? I thought we have it off. Feel free to drop if this is true.

File: src/rust/dns/lib.rs
Comment (mikea) [inline src/rust/dns/lib.rs:49]: you don't need map_err if you have thiserror with `from` setup correctly.

**Review by mikea — APPROVED** [review]
only one issue. otherwise rust side lg.

File: src/rust/dns/lib.rs
Comment (anonrig) [reply src/rust/dns/lib.rs:40]: Yes, if the function is pub then it requires it.

File: src/rust/dns/lib.rs
Comment (anonrig) [reply src/rust/dns/lib.rs:49]: I didn't want to add a specific error for utf16 encoding. invalid hex string with appropriate error message already points us towards this particular line.

**Comment by irvinebroque** [comment]
dev docs?

**Pull request merged**
